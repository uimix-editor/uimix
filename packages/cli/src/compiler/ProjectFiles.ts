import {
  Image,
  ColorToken,
  NodeJSON,
  PageJSON,
  ProjectJSON,
  ProjectManifestJSON,
  StyleJSON,
} from "@uimix/node-data";
import { filter, omit } from "lodash-es";
import { mkdirpSync } from "mkdirp";
import { globSync } from "glob";
import path from "path";
import fs from "fs";
import { sha256 } from "js-sha256";
import { formatJSON } from "../format";
import chokidar from "chokidar";
import { encode } from "url-safe-base64";

function getURLSafeBase64Hash(data: string | Uint8Array): string {
  const hash = sha256.arrayBuffer(data);
  return encode(Buffer.from(hash).toString("base64"));
}

interface HierarchicalNodeJSON extends NodeJSON {
  id: string;
  children: HierarchicalNodeJSON[];
}

function toHierarchicalNodeJSONs(
  nodes: Record<string, NodeJSON>
): HierarchicalNodeJSON[] {
  const hierarchicalNodes: Record<string, HierarchicalNodeJSON> = {};

  for (const [id, node] of Object.entries(nodes)) {
    hierarchicalNodes[id] = {
      ...node,
      id,
      children: [],
    };
  }

  for (const [id, node] of Object.entries(nodes)) {
    if (node.parent) {
      hierarchicalNodes[node.parent].children.push(hierarchicalNodes[id]);
    }
  }

  for (const node of Object.values(hierarchicalNodes)) {
    node.children.sort((a, b) => a.index - b.index);
  }

  return Object.values(hierarchicalNodes).filter((node) => !node.parent);
}

export class ProjectFiles {
  constructor(
    rootPath: string,
    options: {
      filePattern?: string;
    } = {}
  ) {
    this.rootPath = rootPath;
    this.filePattern = options.filePattern ?? "**/*.uimix";
  }

  readonly rootPath: string;
  readonly filePattern: string;
  manifest: ProjectManifestJSON = {};
  pages = new Map<string /* file path */, PageJSON>();

  toProjectJSON(): ProjectJSON {
    // TODO: detect ID conflicts between files (and provide way to resolve them)

    const nodes: Record<string, NodeJSON> = {};
    const styles: Record<string, Partial<StyleJSON>> = {};
    const images: Record<string, Image> = {};
    const colors: Record<string, ColorToken> = {};

    const projectNode: NodeJSON = {
      type: "project",
      index: 0,
    };
    nodes["project"] = projectNode;

    let pageIndex = 0;
    for (const [pageName, pageJSON] of this.pages) {
      const pageID: string = getURLSafeBase64Hash(pageName);

      const pageNode: NodeJSON = {
        type: "page",
        index: pageIndex++,
        name: pageName,
        parent: "project",
      };
      nodes[pageID] = pageNode;

      for (const [id, node] of Object.entries(pageJSON.nodes)) {
        nodes[id] = {
          ...node,
          parent: node.parent ?? pageID,
        };
      }

      for (const [id, style] of Object.entries(pageJSON.styles)) {
        styles[id] = style;
      }
      for (const [id, image] of Object.entries(pageJSON.images ?? {})) {
        images[id] = image;
      }
      for (const [id, color] of Object.entries(pageJSON.colors ?? {})) {
        colors[id] = {
          ...color,
          page: pageID,
        };
      }
    }

    return {
      nodes,
      styles,
      componentURLs: this.manifest.componentURLs,
      images,
      colors,
    };
  }

  loadProjectJSON(projectJSON: ProjectJSON): void {
    const manifest: ProjectManifestJSON = {
      componentURLs: projectJSON.componentURLs,
    };

    const hierarchicalNodes = toHierarchicalNodeJSONs(projectJSON.nodes);
    const projectNode = hierarchicalNodes.find(
      (node) => node.type === "project"
    );
    if (!projectNode) {
      throw new Error("Project node not found");
    }

    const pages = new Map<string, PageJSON>();
    const nodeIDToPageJSON = new Map<string, PageJSON>();
    for (const page of projectNode.children) {
      if (page.type !== "page") {
        throw new Error("expected page node");
      }
      if (!page.name) {
        throw new Error("page name is empty");
      }

      const pageJSON: PageJSON = {
        nodes: {},
        styles: {},
        // TODO: selectively save images
        images: projectJSON.images,
        colors: Object.fromEntries(
          Object.entries(projectJSON.colors ?? {})
            .filter(([, color]) => color.page === page.id)
            .map(([id, color]) => [id, omit(color, ["page"])])
        ),
      };
      const addNodeRecursively = (node: HierarchicalNodeJSON) => {
        pageJSON.nodes[node.id] = omit(node, ["children", "id"]);
        nodeIDToPageJSON.set(node.id, pageJSON);
        for (const child of node.children) {
          addNodeRecursively(child);
        }
      };
      for (const child of page.children) {
        addNodeRecursively(omit(child, ["parent"]));
      }

      pages.set(page.name, pageJSON);
    }

    for (const [id, style] of Object.entries(projectJSON.styles)) {
      const idPath = id.split(":");
      const pageJSON = nodeIDToPageJSON.get(idPath[0]);
      if (pageJSON) {
        pageJSON.styles[id] = style;
      }
    }

    this.manifest = manifest;
    this.pages = pages;
  }

  load(): void {
    const rootPath = this.rootPath;
    const manifestPath = path.resolve(rootPath, "uimix/project.json");
    let manifest: ProjectManifestJSON;
    try {
      manifest = ProjectManifestJSON.parse(
        JSON.parse(fs.readFileSync(manifestPath, { encoding: "utf-8" }))
      );
    } catch {
      manifest = { componentURLs: [] };
    }

    const pages = new Map<string, PageJSON>();

    const pagePaths = globSync(this.filePattern, {
      cwd: rootPath,
    });
    pagePaths.sort();

    for (const pagePath of pagePaths) {
      const pageJSON = PageJSON.parse(
        JSON.parse(
          fs.readFileSync(path.resolve(rootPath, pagePath), {
            encoding: "utf-8",
          })
        )
      );
      pages.set(pagePath.replace(/\.uimix$/, ""), pageJSON);
    }

    this.manifest = manifest;
    this.pages = pages;
  }

  save(): void {
    const rootPath = this.rootPath;

    mkdirpSync(path.resolve(rootPath, "uimix"));
    mkdirpSync(path.resolve(rootPath, "uimix/images"));

    fs.writeFileSync(
      path.resolve(rootPath, "uimix/project.json"),
      formatJSON(JSON.stringify(this.manifest))
    );

    // for (const [hash, image] of Object.entries(this.manifest.images ?? {})) {
    //   const dataURL = image.url;
    //   // data url to buffer
    //   const base64 = dataURL.split(",")[1];
    //   const buffer = Buffer.from(base64, "base64");
    //   const ext = mime.extension(image.type);
    //   if (!ext) {
    //     console.error(`Unknown MIME type for ${image.type}`);
    //     continue;
    //   }

    //   fs.writeFileSync(
    //     path.resolve(rootPath, `uimix/images/${hash}.${ext}`),
    //     buffer
    //   );
    // }

    const pagePathsToDelete = new Set(
      globSync(this.filePattern, {
        cwd: rootPath,
      })
    );

    for (const [pageName, pageJSON] of this.pages) {
      const pagePath = pageName + ".uimix";
      const pageDirPath = path.dirname(pagePath);
      mkdirpSync(path.resolve(rootPath, pageDirPath));
      fs.writeFileSync(
        path.resolve(rootPath, pagePath),
        formatJSON(JSON.stringify(pageJSON))
      );
      pagePathsToDelete.delete(pagePath);
    }

    for (const pagePath of pagePathsToDelete) {
      fs.rmSync(path.resolve(rootPath, pagePath));
    }
  }

  watch(onChange: () => void): () => void {
    const watchPath = path.resolve(this.rootPath, this.filePattern);

    // FIXME: chokidar looks like making UI slow
    const watcher = chokidar.watch(watchPath, {
      ignored: ["**/node_modules/**", "**/dist/**"],
    });
    console.log("start watching...");

    const _onChange = () => {
      try {
        this.load();
        onChange();
      } catch (e) {
        console.error(e);
      }
    };

    watcher.on("change", _onChange);
    watcher.on("add", _onChange);
    watcher.on("unlink", _onChange);

    return () => watcher.close();
  }
}
