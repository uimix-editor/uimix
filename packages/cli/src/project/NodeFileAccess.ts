import * as fs from "fs";
import * as path from "path";
import { FileAccess, Stats } from "./FileAccess";
import chokidar from "chokidar";
import { mkdirp } from "mkdirp";
import { globby } from "globby";

export class NodeFileAccess implements FileAccess {
  watch(cwd: string, patterns: string[], onChange: () => void): () => void {
    const watcher = chokidar.watch(`**/{${patterns.join(",")},}`, {
      cwd,
      ignored: ["**/node_modules/**", "**/.git/**"],
    });
    watcher.on("change", onChange);
    watcher.on("add", onChange);
    watcher.on("unlink", onChange);
    return () => watcher.close();
  }

  async glob(cwd: string, patterns: string[]): Promise<string[]> {
    return (
      await globby(`**/{${patterns.join(",")},}`, {
        cwd,
        ignore: ["**/node_modules/**", "**/.git/**"],
      })
    ).map((filePath) => path.resolve(cwd, filePath));
  }

  async stat(filePath: string): Promise<Stats | undefined> {
    try {
      const stat = await fs.promises.stat(filePath);
      return {
        type: stat.isDirectory() ? "directory" : "file",
      };
    } catch (e) {
      return undefined;
    }
  }

  async writeFile(filePath: string, data: Buffer): Promise<void> {
    await mkdirp(path.dirname(filePath));
    await fs.promises.writeFile(filePath, data);
  }

  async readFile(filePath: string): Promise<Buffer> {
    return await fs.promises.readFile(filePath);
  }

  async remove(filePath: string): Promise<void> {
    await fs.promises.rm(filePath);
  }
}
