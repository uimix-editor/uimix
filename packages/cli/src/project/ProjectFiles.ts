import {
  Image,
  ColorToken,
  NodeJSON,
  PageJSON,
  ProjectJSON,
  ProjectManifestJSON,
  StyleJSON,
} from "@uimix/model/src/data/v1";
import { getPageID, compareProjectJSONs } from "@uimix/model/src/data/util";
import { omit } from "lodash-es";
import { formatJSON } from "../format";
import { FileAccess } from "./FileAccess";

interface HierarchicalNodeJSON extends NodeJSON {
  id: string;
  children: HierarchicalNodeJSON[];
}

function toHierarchicalNodeJSONs(
  nodes: Record<string, NodeJSON>
): HierarchicalNodeJSON[] {
  const hierarchicalNodes: Record<string, HierarchicalNodeJSON> = {};

  for (const [id, node] of Object.entries(nodes)) {
    hierarchicalNodes[id] = {
      ...node,
      id,
      children: [],
    };
  }

  for (const [id, node] of Object.entries(nodes)) {
    if (node.parent) {
      hierarchicalNodes[node.parent].children.push(hierarchicalNodes[id]);
    }
  }

  for (const node of Object.values(hierarchicalNodes)) {
    node.children.sort((a, b) => a.index - b.index);
  }

  return Object.values(hierarchicalNodes).filter((node) => !node.parent);
}

interface ProjectFilesOptions {
  filePattern?: string;
}

// Important TODO: fix paths in Windows!!
export class ProjectFiles {
  static async load(fileAccess: FileAccess, options: ProjectFilesOptions = {}) {
    const projectFiles = new ProjectFiles(fileAccess, options);
    await projectFiles.load();
    return projectFiles;
  }

  constructor(fileAccess: FileAccess, options: ProjectFilesOptions = {}) {
    this.fileAccess = fileAccess;
    this.filePattern = options.filePattern ?? "**/*.uimix";
  }

  readonly fileAccess: FileAccess;

  get rootPath(): string {
    return this.fileAccess.rootPath;
  }

  readonly filePattern: string;
  readonly manifestName = "uimix.json";
  json: ProjectJSON = {
    nodes: {},
    styles: {},
  };

  async load(): Promise<boolean> {
    let manifest: ProjectManifestJSON;
    try {
      manifest = ProjectManifestJSON.parse(
        JSON.parse(await this.fileAccess.readText(this.manifestName))
      );
    } catch {
      manifest = { componentURLs: [] };
    }

    const pages = new Map<string, PageJSON>();

    const pagePaths = await this.fileAccess.glob(this.filePattern);
    pagePaths.sort();

    for (const pagePath of pagePaths) {
      const pageJSON = PageJSON.parse(
        JSON.parse(await this.fileAccess.readText(pagePath))
      );
      pages.set(pagePath.replace(/\.uimix$/, ""), pageJSON);
    }
    const newProjectJSON = filesToProjectJSON(manifest, pages);

    if (compareProjectJSONs(this.json, newProjectJSON)) {
      return false;
    }

    this.json = newProjectJSON;
    return true;
  }

  async save(): Promise<void> {
    const { manifest, pages } = projectJSONToFiles(this.json);

    await this.fileAccess.writeText(
      this.manifestName,
      formatJSON(JSON.stringify(manifest))
    );

    const pagePathsToDelete = new Set(
      await this.fileAccess.glob(this.filePattern)
    );

    for (const [pageName, pageJSON] of pages) {
      const pagePath = pageName + ".uimix";
      await this.fileAccess.writeText(
        pagePath,
        formatJSON(JSON.stringify(pageJSON))
      );
      pagePathsToDelete.delete(pagePath);
    }

    for (const pagePath of pagePathsToDelete) {
      await this.fileAccess.remove(pagePath);
    }
  }

  watch(onChange: (projectJSON: ProjectJSON) => void): () => void {
    console.log("start watching...");

    return this.fileAccess.watch(this.filePattern, async () => {
      try {
        if (await this.load()) {
          onChange(this.json);
        }
      } catch (e) {
        console.error(e);
      }
    });
  }
}

export function projectJSONToFiles(projectJSON: ProjectJSON): {
  manifest: ProjectManifestJSON;
  pages: Map<string, PageJSON>;
} {
  const manifest: ProjectManifestJSON = {
    componentURLs: projectJSON.componentURLs,
  };

  const hierarchicalNodes = toHierarchicalNodeJSONs(projectJSON.nodes);
  const projectNode = hierarchicalNodes.find((node) => node.type === "project");
  if (!projectNode) {
    throw new Error("Project node not found");
  }

  const pages = new Map<string, PageJSON>();
  const nodeIDToPageJSON = new Map<string, PageJSON>();
  for (const page of projectNode.children) {
    if (page.type !== "page") {
      throw new Error("expected page node");
    }
    if (!page.name) {
      throw new Error("page name is empty");
    }

    const pageJSON: PageJSON = {
      nodes: {},
      styles: {},
      // TODO: selectively save images
      images: projectJSON.images,
      colors: Object.fromEntries(
        Object.entries(projectJSON.colors ?? {})
          .filter(([, color]) => color.page === page.id)
          .map(([id, color]) => [id, omit(color, ["page"])])
      ),
    };
    const addNodeRecursively = (node: HierarchicalNodeJSON) => {
      pageJSON.nodes[node.id] = omit(node, ["children", "id"]);
      nodeIDToPageJSON.set(node.id, pageJSON);
      for (const child of node.children) {
        addNodeRecursively(child);
      }
    };
    for (const child of page.children) {
      addNodeRecursively(omit(child, ["parent"]));
    }

    pages.set(page.name, pageJSON);
  }

  for (const [id, style] of Object.entries(projectJSON.styles)) {
    const idPath = id.split(":");
    const pageJSON = nodeIDToPageJSON.get(idPath[0]);
    if (pageJSON) {
      pageJSON.styles[id] = style;
    }
  }

  return {
    manifest,
    pages,
  };
}

export function filesToProjectJSON(
  manifest: ProjectManifestJSON,
  pages: Map<string, PageJSON>
): ProjectJSON {
  // TODO: detect ID conflicts between files (and provide way to resolve them)

  const nodes: Record<string, NodeJSON> = {};
  const styles: Record<string, Partial<StyleJSON>> = {};
  const images: Record<string, Image> = {};
  const colors: Record<string, ColorToken> = {};

  const projectNode: NodeJSON = {
    type: "project",
    index: 0,
  };
  nodes["project"] = projectNode;

  let pageIndex = 0;
  for (const [pageName, pageJSON] of pages) {
    const pageID: string = getPageID(pageName);

    const pageNode: NodeJSON = {
      type: "page",
      index: pageIndex++,
      name: pageName,
      parent: "project",
    };
    nodes[pageID] = pageNode;

    for (const [id, node] of Object.entries(pageJSON.nodes)) {
      nodes[id] = {
        ...node,
        parent: node.parent ?? pageID,
      };
    }

    for (const [id, style] of Object.entries(pageJSON.styles)) {
      styles[id] = style;
    }
    for (const [id, image] of Object.entries(pageJSON.images ?? {})) {
      images[id] = image;
    }
    for (const [id, color] of Object.entries(pageJSON.colors ?? {})) {
      colors[id] = {
        ...color,
        page: pageID,
      };
    }
  }

  return {
    nodes,
    styles,
    componentURLs: manifest.componentURLs,
    images,
    colors,
  };
}
