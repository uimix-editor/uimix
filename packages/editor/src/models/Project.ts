/* eslint-disable @typescript-eslint/no-explicit-any */

import * as Y from "yjs";
import { posix as path } from "path-browserify";
import { getOrCreate } from "../state/Collection";
import { SelectableMap } from "./Selectable";
import { Node, NodeMap } from "./Node";
import { computed, makeObservable, ObservableMap } from "mobx";
import { ProjectJSON } from "@uimix/node-data";
import { toProjectJSON } from "./toProjectJSON";
import { ImageManager } from "./ImageManager";
import { Component } from "./Component";
import { ObservableYMap } from "../utils/ObservableYMap";

export interface PageHierarchyFolderEntry {
  type: "directory";
  path: string;
  name: string;
  children: PageHierarchyEntry[];
}

export interface PageHierarchyPageEntry {
  type: "file";
  path: string;
  name: string;
  page: Node;
}

export type PageHierarchyEntry =
  | PageHierarchyFolderEntry
  | PageHierarchyPageEntry;

class Pages {
  constructor(project: Project) {
    this.project = project;
    this.node = project.node;
    makeObservable(this);
  }

  node: Node;
  readonly project: Project;

  @computed get count(): number {
    return this.node.childCount;
  }

  get all(): Node[] {
    return this.node.children;
  }

  create(filePath: string): Node {
    const node = this.project.nodes.create("page");
    node.name = filePath;
    return node;
  }

  toHierarchy(): PageHierarchyFolderEntry {
    const root: PageHierarchyFolderEntry = {
      type: "directory",
      name: "",
      path: "",
      children: [],
    };
    const parents = new Map<string, PageHierarchyFolderEntry>();
    parents.set("", root);

    const mkdirp = (segments: string[]): PageHierarchyFolderEntry => {
      if (segments.length === 0) {
        return root;
      }

      const existing = parents.get(segments.join("/"));
      if (existing) {
        return existing;
      }

      const parent = mkdirp(segments.slice(0, -1));
      const dir: PageHierarchyFolderEntry = {
        type: "directory",
        name: segments[segments.length - 1],
        path: segments.join("/"),
        children: [],
      };
      parent.children.push(dir);
      parents.set(segments.join("/"), dir);
      return dir;
    };

    const pages = Array.from(this.all);
    pages.sort((a, b) => a.name.localeCompare(b.name));

    for (const page of pages) {
      const segments = page.name.split(path.sep);
      const parent = mkdirp(segments.slice(0, -1));

      const item: PageHierarchyPageEntry = {
        type: "file",
        name: segments[segments.length - 1],
        path: page.name,
        page,
      };
      parent.children.push(item);
    }

    return root;
  }

  affectedPagesForPath(path: string): Node[] {
    return this.all.filter(
      (page) => page.name === path || page.name.startsWith(path + "/")
    );
  }

  deletePageOrPageFolder(path: string) {
    const pagesToDelete = this.affectedPagesForPath(path);

    for (const page of pagesToDelete) {
      page.remove();
      // TODO: delete dangling nodes?
      //this.project.nodes.remove(doc.root);
    }
  }

  renamePageOrPageFolder(path: string, newPath: string) {
    if (path === newPath) {
      return;
    }

    const pagesToDelete = this.affectedPagesForPath(path);

    for (const page of pagesToDelete) {
      page.name = newPath + page.name.slice(path.length);
    }
  }
}

export class Project {
  constructor(data: Y.Map<any>) {
    this.data = ObservableYMap.get(data);
    this.nodes = new NodeMap(this);
    this.selectables = new SelectableMap(this);
    this.node = this.nodes.create("project", "project");
    this.pages = new Pages(this);
  }

  readonly data: ObservableYMap<any>;
  readonly nodes: NodeMap;
  readonly selectables: SelectableMap;
  readonly node: Node;
  readonly pages: Pages;
  readonly imageManager = new ImageManager();

  toJSON(): ProjectJSON {
    return toProjectJSON(this.node.children.map((c) => c.selectable));
  }

  loadJSON(json: ProjectJSON) {
    this.node.clear();

    for (const [id, nodeJSON] of Object.entries(json.nodes)) {
      const node = this.nodes.getOrCreate(nodeJSON.type, id);
      node.loadJSON(nodeJSON);
      if (node.type === "page") {
        this.node.append([node]);
      }
    }
    for (const [id, style] of Object.entries(json.styles)) {
      const selectable = this.selectables.get(id.split(":"));
      selectable.selfStyle.loadJSON(style);
    }
  }

  get components(): Component[] {
    const components: Component[] = [];

    for (const page of this.pages.all) {
      for (const node of page.children) {
        const component = Component.from(node);
        if (component) {
          components.push(component);
        }
      }
    }

    return components;
  }
}
