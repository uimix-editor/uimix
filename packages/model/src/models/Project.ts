import * as Y from "yjs";
import { Selectable, SelectableMap } from "./Selectable";
import { Node, NodeMap } from "./Node";
import * as Data from "../data/v1";
import { ImageManager } from "./ImageManager";
import { Component } from "./Component";
import { ObservableYArray } from "@uimix/foundation/src/utils/ObservableYArray";
import { ColorTokenMap } from "./ColorToken";
import { PageList } from "./PageList";
import { ProjectData } from "../collaborative/ProjectData";
import { CodeAssets } from "./CodeAssets";
import { makeObservable, observable } from "mobx";
import { Page } from "./Page";

export class Project {
  constructor() {
    this.nodes = new NodeMap(this);
    this.selectables = new SelectableMap(this);
    this.colorTokens = new ColorTokenMap(this);
    this.node = this.nodes.create("project", "project");
    this.pages = new PageList(this);
    makeObservable(this);
  }

  readonly data = new ProjectData();
  readonly nodes: NodeMap;
  readonly selectables: SelectableMap;
  readonly colorTokens: ColorTokenMap;
  readonly node: Node;
  readonly pages: PageList;
  readonly imageManager = new ImageManager(this);

  get doc(): Y.Doc {
    return this.data.doc;
  }

  createUndoManager(): Y.UndoManager {
    return new Y.UndoManager([
      this.data.nodes,
      this.data.styles,
      this.data.selection,
      this.data.componentURLs,
      this.data.colors,
    ]);
  }

  clear(): void {
    this.data.clear();
    this.nodes.create("project", "project");
  }

  toJSON(): Data.Project {
    return this.data.toJSON();
  }

  loadJSON(json: Data.Project) {
    this.data.loadJSON(json);
  }

  get components(): Component[] {
    return this.pages.all.flatMap((page) => page.components);
  }

  get componentURLs(): ObservableYArray<string> {
    return ObservableYArray.get(this.data.componentURLs);
  }

  @observable localCodeAssets: CodeAssets | undefined = undefined;

  // Selection

  clearSelection() {
    this.selectables.selectionData.clear();
  }

  replaceSelection(selectables: Iterable<Selectable>) {
    this.clearSelection();
    for (const selectable of selectables) {
      selectable.select();
    }
  }

  // Get by ID

  nodeForID(id: string): Node | undefined {
    return this.nodes.get(id);
  }

  pageForID(id: string): Page | undefined {
    const node = this.nodeForID(id);
    if (node) {
      return Page.from(node);
    }
  }

  componentForID(id: string): Component | undefined {
    const node = this.nodeForID(id);
    if (node) {
      return Component.from(node);
    }
  }
}
