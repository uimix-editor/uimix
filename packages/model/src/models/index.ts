export * from "./ColorRef";
export * from "./ColorToken";
export * from "./Component";
export * from "./ImageManager";
export * from "./Node";
export * from "./ObjectData";
export * from "./Page";
export * from "./PageList";
export * from "./Project";
export * from "../collaborative/ProjectData";
export * from "./Selectable";
export * from "./Style";
