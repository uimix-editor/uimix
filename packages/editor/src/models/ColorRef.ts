import { Color } from "../utils/Color";
import { ColorToken } from "./ColorToken";
import { Project } from "./Project";
import { Color as ColorJSON } from "@uimix/node-data";

export class ColorRef {
  constructor(value: Color | ColorToken) {
    this.value = value;
  }

  static fromJSON(project: Project, json: ColorJSON): ColorRef | undefined {
    if (typeof json === "string") {
      const color = Color.from(json);
      return color && new ColorRef(color);
    }
    const token = project.colorTokens.get(json.id);
    return token && new ColorRef(token);
  }

  toJSON(): ColorJSON {
    if (this.value instanceof Color) {
      return this.value.toHex();
    } else {
      return {
        type: "token",
        id: this.value.id,
      };
    }
  }

  get color(): Color {
    if (this.value instanceof Color) {
      return this.value;
    } else {
      return this.value.value ?? Color.black;
    }
  }

  readonly value: Color | ColorToken;
}
