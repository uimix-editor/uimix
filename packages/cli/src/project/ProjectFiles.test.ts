import { describe, it, expect, beforeEach, afterEach } from "vitest";
import { ProjectFiles } from "./ProjectFiles";
import * as fs from "fs";
import shell from "shelljs";
import tmp from "tmp";
import { Project } from "@uimix/model/src/models/Project";
import { NodeFileAccess } from "./NodeFileAccess";

describe(ProjectFiles.name, () => {
  let tmpObj: tmp.DirResult;

  beforeEach(async () => {
    tmpObj = tmp.dirSync({
      unsafeCleanup: true,
    });
    shell.mkdir("-p", tmpObj.name + "/demo-project");
  });

  afterEach(async () => {
    tmpObj.removeCallback();
  });

  // it("loads", () => {
  //   const projectFiles = new ProjectFiles(tmpObj.name + "/demo-project");
  //   projectFiles.load();

  //   expect(projectFiles.toProjectJSON()).toMatchSnapshot();
  // });

  it("saves", async () => {
    const project = new Project();
    const page1 = project.pages.create("src/page1");
    const page2 = project.pages.create("src/page2");

    page1.node.append([project.nodes.create("frame", "frame1")]);
    page2.node.append([project.nodes.create("text", "text1")]);

    const projectFiles = new ProjectFiles(
      new NodeFileAccess(tmpObj.name + "/demo-project")
    );
    projectFiles.json = project.toJSON();
    await projectFiles.save();

    const page1File = fs.readFileSync(
      tmpObj.name + "/demo-project/src/page1.uimix",
      "utf8"
    );
    expect(page1File).toMatchSnapshot();

    const page2File = fs.readFileSync(
      tmpObj.name + "/demo-project/src/page2.uimix",
      "utf8"
    );
    expect(page2File).toMatchSnapshot();

    const projectFiles2 = await ProjectFiles.load(
      new NodeFileAccess(tmpObj.name + "/demo-project")
    );

    expect(projectFiles2.json).toEqual(project.toJSON());
  });

  it("watches", async () => {
    const project = new Project();
    const page1 = project.pages.create("src/page1");
    const page2 = project.pages.create("src/page2");

    page1.node.append([project.nodes.create("frame", "frame1")]);
    page2.node.append([project.nodes.create("text", "text1")]);

    const projectFiles = new ProjectFiles(
      new NodeFileAccess(tmpObj.name + "/demo-project")
    );
    projectFiles.json = project.toJSON();

    let watchCount = 0;
    projectFiles.watch(() => {
      watchCount++;
    });
    await new Promise((resolve) => setTimeout(resolve, 500));

    // saves not cause watch
    await projectFiles.save();
    await new Promise((resolve) => setTimeout(resolve, 500));
    expect(watchCount).toBe(0);

    // change file
    fs.rmSync(tmpObj.name + "/demo-project/src/page1.uimix");
    await new Promise((resolve) => setTimeout(resolve, 500));

    expect(watchCount).toBe(1);

    project.loadJSON(projectFiles.json);

    expect(project.pages.all.map((page) => page.filePath)).toEqual([
      "src/page2",
    ]);
  });
});
