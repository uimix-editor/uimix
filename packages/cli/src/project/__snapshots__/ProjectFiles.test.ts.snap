// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProjectFiles > saves 1`] = `
"{
  \\"nodes\\": { \\"frame1\\": { \\"type\\": \\"frame\\", \\"index\\": 0 } },
  \\"styles\\": {},
  \\"images\\": {},
  \\"colors\\": {}
}
"
`;

exports[`ProjectFiles > saves 2`] = `
"{
  \\"nodes\\": { \\"text1\\": { \\"type\\": \\"text\\", \\"index\\": 0 } },
  \\"styles\\": {},
  \\"images\\": {},
  \\"colors\\": {}
}
"
`;
